src/IO/AsciiPCLoader/asciipcloader.cpp

```diff
@@ -0,0 +1,195 @@
+
+    if ( ( xpos == -1 ) || ( ypos == -1 ) || ( zpos == -1 ) || ( timepos == -1 ) ) {
+        delete fileData;
+        LOG( logERROR ) << "file does not contain mandatory properties (X, Y, Z, Time). aborting";
+        return nullptr;
+    }
+    else {
+        LOG( logINFO ) << "loading properties: ";
+        string info = "";
+        for ( size_t i = 0; i < tokens.size(); ++i ) {
+            if ( i % 4 == 3 ) {
+                LOG( logINFO ) << info << tokens[i];
+                info = "";
+            }
+            else {
+                info += tokens[i] + ", ";
+            }
+        }
+    }
+
+    Ra::Core::VectorArray<Ra::Core::Vector3>& vertices = geometry->getGeometry().verticesWithLock();
+
+    // creating custom attrib for time record
+    auto handle  = geometry->getGeometry().vertexAttribs().addAttrib<Scalar>( "time" );
+    auto& attrib = geometry->getGeometry().vertexAttribs().getAttrib( handle );
+    auto& time   = attrib.getDataWithLock();
+
+    // creating custom attribs for all other properties
+    vector<Ra::Core::VectorArray<Scalar>> attribs;
+    for ( size_t i = 0; i < tokens.size(); ++i ) {
+        // skipping mandatory attribs
+        if ( ( (int)i == xpos ) || ( (int)i == ypos ) || ( (int)i == zpos ) ||
+             ( (int)i == timepos ) ) {
+            continue;
+        }
+        auto handle2  = geometry->getGeometry().vertexAttribs().addAttrib<Scalar>( tokens[i] );
+        auto& attrib2 = geometry->getGeometry().vertexAttribs().getAttrib( handle2 );
+        attribs.emplace_back( attrib2.getDataWithLock() );
+    }
+
+    /*recording data*/
+    vector<double> pointRecord;
+    int i = 0;
+    double num;
+    while ( stream >> num ) {
+        pointRecord.emplace_back( num );
+
+        if ( i == (int)( tokens.size() - 1 ) ) {
+            vertices.emplace_back( Scalar( pointRecord[xpos] ),
+                                   Scalar( pointRecord[ypos] ),
+                                   Scalar( pointRecord[zpos] ) );
+            time.emplace_back( Scalar( pointRecord[timepos] ) );
+            size_t k = 0;
+            for ( size_t j = 0; j < tokens.size(); ++j ) {
+                if ( ( (int)j == xpos ) || ( (int)j == ypos ) || ( (int)j == zpos ) ||
+                     ( (int)j == timepos ) ) {
+                    continue;
+                }
+                attribs[k].emplace_back( Scalar( pointRecord[j] ) );
+                ++k;
+            }
+            pointRecord.clear();
+        }
+        i = ( i + 1 ) % tokens.size();
+    }
+    stream.close();
+
+    geometry->getGeometry().verticesUnlock();
+    geometry->getGeometry().vertexAttribs().unlock( handle );
+    for ( size_t i = 0; i < tokens.size(); ++i ) {
+        if ( ( (int)i == xpos ) || ( (int)i == ypos ) || ( (int)i == zpos ) ||
+             ( (int)i == timepos ) ) {
+            continue;
+        }
+        auto handle2 = geometry->getGeometry().vertexAttribs().findAttrib<Scalar>( tokens[i] );
```

Comment:
I guess it would be better to store the handle in an array, it would be easier to lock/unlock.

Reply:
updated